test.py

```diff
@@ -0,0 +1,595 @@
+                func = start_test_background
+                args = framework, avail_cluster, repo_root
+                _action_wrapper("Launch %s tests" % framework.name,
+                        framework, func, *args)
+                next_test = None
+                avail_cluster = None
+            else:
+                # No tests left, handle completion and waiting for completion.
+                if not fwinfo.running_frameworks():
+                    logger.info("No framework tests running.  All done.")
+                    all_ok = True
+                    break # all tests done
+                logger.info("No framework tests to launch, waiting for completions.")
+                # echo status
+                time.sleep(30) # waiting for tests to complete
+
+            # after launching a test, or waiting, check for test completion.
+            all_ok = _handle_test_completions()
+            if not all_ok:
+                logger.info("Some tests failed; aborting early") # TODO paramaterize
+                break
+    finally:
+        # TODO probably should also make this teardown optional
+        for framework_name in fwinfo.get_framework_names():
+            logger.info("Terminating subprocess for framework=%s", framework_name)
+            framework = fwinfo.get_framework(framework_name)
+            if framework.popen:
+                framework.popen.terminate() # does nothing if already completed
+    return all_ok
+
+def run_tests(run_attrs, repo_root):
+    logger.info("cluster_teardown policy: %s", run_attrs.cluster_teardown)
+    try: # all clusters are set up inside this try
+        all_passed = False
+        if run_attrs.parallel:
+            logger.debug("Running m ulticluster test run")
+            all_passed = _multicluster_linear_per_cluster(run_attrs, repo_root)
+        else:
+            all_passed = _one_cluster_linear_tests(run_attrs, repo_root)
+    finally:
+        if run_attrs.cluster_teardown == "always":
+            teardown_clusters()
+        elif run_attrs.cluster_teardown == "success-only" and all_passed:
+            teardown_clusters()
+
+        for cluster in clustinfo._clusters:
+            logger.debug("Cluster still running: url=%s id=%s auth_token=%s",
+                         cluster.url, cluster.cluster_id, cluster.auth_token)
+
+
+def _setup_strict(framework, cluster, repo_root):
+    "Do cruft required to create the roles for the framework"
+    security = os.environ.get('SECURITY', '')
+    logger.info("SECURITY set to: '%s'", security)
+    if security == "strict":
+        logger.info("running %s role strict setup script(s)", framework.name)
+        perm_setup_script = os.path.join(repo_root, 'tools', 'setup_permissions.sh')
+
+        custom_env = os.environ.copy()
+        custom_env['CLUSTER_URL'] = cluster.url
+        custom_env['CLUSTER_AUTH_TOKEN'] = cluster.auth_token
+
+        for script_num in (1, 2):
```

Comment:
pre-existing magic numbers.  I wasn't changing any of the "external interfaces" of this while I wrote it because there was more than enough to figure out, and finding the exceptions usually only happens after you've merged the change.

I suspect what this should be a is a declarative file in the framework test definitions saying what roles they want, where this parses and creates them.